llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,225 @@
+  }
+
+  // run the SYCLMutatePrintfAddrspace pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLMutatePrintfAddrspacePass Impl;
+};
+
+struct CallReplacer;
+class AddrspaceReplacer {
+public:
+  static constexpr unsigned ConstantAddrspaceID = 2;
+
+  AddrspaceReplacer() = delete;
+  AddrspaceReplacer(Module *M) : M(M) {
+    Type *Int8Ty = Type::getInt8Ty(M->getContext());
+    CASLiteralType = PointerType::get(Int8Ty, ConstantAddrspaceID);
+  }
+  ~AddrspaceReplacer() {
+    for (Function *F : FunctionsToDrop)
+      F->eraseFromParent();
+  }
+  size_t getReplacedCallsCount() { return ReplacedCallsCount; }
+
+  void runOnFunctionDeclaration(Function *F);
+
+private:
+  Module *M;
+  PointerType *CASLiteralType;
+  size_t ReplacedCallsCount = 0;
+  /// If the variadic version gets picked during FE compilation, we'll only have
+  /// 1 function to replace. However, unique declarations are emitted for each
+  /// of the non-variadic (variadic template) calls.
+  SmallVector<Function *, 8> FunctionsToDrop;
+
+  Function *getCASPrintfFunction(Function *GenericASPrintfFunc);
+  Constant *getCASLiteral(GlobalVariable *GenericASLiteral);
+  CallReplacer prepareCASArgForCall(CallInst *CI);
+};
+} // namespace
+
+char SYCLMutatePrintfAddrspaceLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLMutatePrintfAddrspaceLegacyPass,
+                "SYCLMutatePrintfAddrspace",
+                "Move SYCL printf literal arguments to constant address space",
+                false, false)
+
+// Public interface to the SYCLMutatePrintfAddrspacePass.
+ModulePass *llvm::createSYCLMutatePrintfAddrspaceLegacyPass() {
+  return new SYCLMutatePrintfAddrspaceLegacyPass();
+}
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  AddrspaceReplacer ASReplacer(&M);
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+    if (!F.getName().startswith("_Z18__spirv_ocl_printf"))
+      continue;
+    ASReplacer.runOnFunctionDeclaration(&F);
+  }
+  return ASReplacer.getReplacedCallsCount() ? PreservedAnalyses::all()
+                                            : PreservedAnalyses::none();
+}
+
+/// Helper implementations
+namespace {
+
+/// Encapsulates the update of CallInst's literal argument.
+struct CallReplacer {
+  CallReplacer(CallInst *CI, Value *CASArg) : CI(CI), CASArg(CASArg) {}
+  void replaceWithFunction(Function *CASPrintf) {
+    CI->setArgOperand(0, CASArg);
+    CI->setCalledFunction(CASPrintf);
+  }
+
+private:
+  CallInst *CI;
+  Value *CASArg;
+};
+
+void AddrspaceReplacer::runOnFunctionDeclaration(Function *F) {
+  auto *LiteralType = F->getArg(0)->getType();
+  if (LiteralType->getPointerAddressSpace() == ConstantAddrspaceID)
+    // No need to replace the literal type and its printf users
+    return;
+  Function *CASPrintfFunc = getCASPrintfFunction(F);
```

Comment:
We do (hence `Module::getOrInsertFunction()`, and technically, we can construct it once per each `Module`. Additionally, there's not much need for the `Function *` argument, since we can get the needed function type from `Context` alone. I'll adjust accordingly.